### PR TITLE
Sending works both with tls and zlib compression

### PR DESCRIPTION
**Pull request by ludwikbukowski**

### CODE REVIEW

File: src/escalus_tcp.erl [hunk 1]
```diff
@@ -227,6 +227,9 @@ handle_call(stop, _From, #state{} = S) ->
 handle_cast({send, #client{socket = Socket, ssl = Ssl, compress = Compress},
              Elem}, #state{on_request = OnRequestFun} = State) ->
     Reply = case {Ssl, Compress} of
+                {true, {zlib, {_,Zout}}} ->
+                    Deflated = zlib:deflate(Zout, exml:to_iolist(Elem), sync),
+                    ssl:send(Socket, Deflated);
                 {true, _} ->
                     ssl:send(Socket, exml:to_iolist(Elem));
                 {false, {zlib, {_,Zout}}} ->
```
Comment (gadgetci) [inline src/escalus_tcp.erl]: According to **Elvis**:
> Missing space after "," on line 235

File: src/escalus_tcp.erl [hunk 2]
```diff
@@ -227,6 +227,9 @@ handle_call(stop, _From, #state{} = S) ->
 handle_cast({send, #client{socket = Socket, ssl = Ssl, compress = Compress},
              Elem}, #state{on_request = OnRequestFun} = State) ->
     Reply = case {Ssl, Compress} of
+                {true, {zlib, {_,Zout}}} ->
```
Comment (gadgetci) [inline src/escalus_tcp.erl]: According to **Elvis**:
> Missing space after "," on line 230

File: src/escalus_tcp.erl [hunk 3]
```diff
@@ -227,6 +227,9 @@ handle_call(stop, _From, #state{} = S) ->
 handle_cast({send, #client{socket = Socket, ssl = Ssl, compress = Compress},
```
Comment (gadgetci) [inline src/escalus_tcp.erl]: According to **Compiler**:
> Warning: missing specification for function handle_cast/2

**Comment by michalwski** [comment]
@ludwikbukowski, please apply changes suggested by @gadgetci.

File: src/escalus_tcp.erl [hunk 4]
```diff
@@ -224,12 +224,16 @@ handle_call(stop, _From, #state{} = S) ->
     wait_until_closed(S#state.socket),
     {stop, normal, ok, S}.
 
+-spec handle_cast({send, #client{}, any()}, #state{}) -> {noreply, #state{}}.
```
Comment (gadgetci) [inline src/escalus_tcp.erl]: According to **Elvis**:
> The spec in line 227 uses a record, please define a type for the record and use that instead.

**Comment by michalwski** [comment]
This is for esl/MongooseIM#616

File: src/ct_tty_hook.erl [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+pre_init_per_group(Group,Config,State) ->
+  print_group_enter(Group, State, "Starting"),
+  {Config, State}.
+
+%% @doc Called after each init_per_group.
+post_init_per_group(_Group,_Config,Return,State) ->
+  {Return, State}.
+
+%% @doc Called after each end_per_group.
+pre_end_per_group(_Group,Config,State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 72

File: src/ct_tty_hook.erl [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  {Return, State#state{ data = [Data | State#state.data] ,
+    total = State#state.total + State#state.suite_total } }.
+
+%% @doc Called before each init_per_group.
+pre_init_per_group(Group,Config,State) ->
+  print_group_enter(Group, State, "Starting"),
+  {Config, State}.
+
+%% @doc Called after each init_per_group.
+post_init_per_group(_Group,_Config,Return,State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 68

File: src/ct_tty_hook.erl [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+print_suite_ok(Name, OkCount) ->
+  io:format("~n====== Suite OK: ~p (All ~p tests passed)~n",
+    [Name, OkCount]).
+print_suite_failed(Name, OkCount, FailCount, Cases) ->
+  ct:pal("~n====== Suite FAILED: ~p (~p of ~p tests failed)~n",
+    [Name, FailCount, OkCount+FailCount]),
+  lists:foreach(fun maybe_print_test_case/1, Cases).
+
+maybe_print_test_case({testcase, _Name,ok,_})              -> ok;
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 133

File: src/ct_tty_hook.erl [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+%% @doc Called after init_per_suite.
+post_init_per_suite(_Suite,_Config,Return,State) ->
+  {Return, State}.
+
+%% @doc Called before end_per_suite.
+pre_end_per_suite(_Suite,Config,State) ->
+  {Config, State}.
+
+%% @doc Called after end_per_suite.
+post_end_per_suite(Suite,_Config,Return,State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 57

File: src/ct_tty_hook.erl [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+%% any common state.
+init(_Id, Opts) ->
+  Unfolded = proplists:unfold(Opts),
+  PrintGroup = proplists:get_value(print_group, Unfolded, false),
+  PrintCase = proplists:get_value(print_case, Unfolded, false),
+  {ok, #state{ total = 0, data = [],
+    print_group = PrintGroup, print_case = PrintCase }}.
+
+%% @doc Called before init_per_suite is called.
+pre_init_per_suite(Suite,Config,State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 45

File: src/ct_tty_hook.erl [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+%% @doc Called after each init_per_group.
+post_init_per_group(_Group,_Config,Return,State) ->
+  {Return, State}.
+
+%% @doc Called after each end_per_group.
+pre_end_per_group(_Group,Config,State) ->
+  {Config, State}.
+
+%% @doc Called after each end_per_group.
+post_end_per_group(Group,_Config,Return,State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 76

File: src/ct_tty_hook.erl [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+  print_group_enter(Group, State, "Finished"),
+  {Return, State}.
+
+%% @doc Called before each test case.
+pre_init_per_testcase(TC,Config,State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 81

File: src/ct_tty_hook.erl [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+aggregate_results(Cases) ->
+  Oks =   [ C || {testcase, _, ok, _} =         C <- Cases ],
+  Fails = [ C || {testcase, _, {error, _}, _} = C <- Cases ],
+  {length(Oks), length(Fails)}.
+
+print_suite({suites, Name,_, TestCases}) ->
+  case aggregate_results(TestCases) of
+    {Oks, 0} -> print_suite_ok(Name, Oks);
+    {Oks, Fails} -> print_suite_failed(Name,Oks,Fails,TestCases)
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 122

File: src/ct_tty_hook.erl [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+%% @doc Called before init_per_suite is called.
+pre_init_per_suite(Suite,Config,State) ->
+  {Config, State#state{ suite_total = 0, tcs = [] }}.
+
+%% @doc Called after init_per_suite.
+post_init_per_suite(_Suite,_Config,Return,State) ->
+  {Return, State}.
+
+%% @doc Called before end_per_suite.
+pre_end_per_suite(_Suite,Config,State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 53

File: src/ct_tty_hook.erl [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+  PrintCase = proplists:get_value(print_case, Unfolded, false),
+  {ok, #state{ total = 0, data = [],
+    print_group = PrintGroup, print_case = PrintCase }}.
+
+%% @doc Called before init_per_suite is called.
+pre_init_per_suite(Suite,Config,State) ->
+  {Config, State#state{ suite_total = 0, tcs = [] }}.
+
+%% @doc Called after init_per_suite.
+post_init_per_suite(_Suite,_Config,Return,State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 49

File: src/ct_tty_hook.erl (hunk 9 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function pre_init_per_suite/3

File: src/ct_tty_hook.erl [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+  {Config, State}.
+
+%% @doc Called after end_per_suite.
+post_end_per_suite(Suite,_Config,Return,State) ->
+  Data = {suites, Suite, State#state.suite_total, lists:reverse(State#state.tcs)},
+  {Return, State#state{ data = [Data | State#state.data] ,
+    total = State#state.total + State#state.suite_total } }.
+
+%% @doc Called before each init_per_group.
+pre_init_per_group(Group,Config,State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function pre_init_per_group/3

File: src/ct_tty_hook.erl [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+-export([terminate/1]).
+-record(state, { total, suite_total, ts, tcs, data, print_group, print_case }).
+
+%% @doc Return a unique id for this CTH.
+id(_Opts) ->
+  "ct_tty_hook_001".
+
+%% @doc Always called before any other callback function. Use this to initiate
+%% any common state.
+init(_Id, Opts) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function init/2

File: src/ct_tty_hook.erl [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+  print_group_enter(Group, State, "Finished"),
+  {Return, State}.
+
+%% @doc Called before each test case.
+pre_init_per_testcase(TC,Config,State) ->
+  print_case_enter(TC, State, "Starting"),
+  {Config, State#state{ ts = now(), total = State#state.suite_total + 1 } }.
+
+%% @doc Called after each test case.
+post_end_per_testcase(TC,_Config,Return,State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function post_end_per_testcase/4

File: src/ct_tty_hook.erl (hunk 13 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function pre_end_per_suite/3

File: src/ct_tty_hook.erl (hunk 10 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function post_end_per_group/4

File: src/ct_tty_hook.erl (hunk 14 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function post_init_per_suite/4

File: src/ct_tty_hook.erl [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+  TCInfo = {testcase, TC, Return, timer:now_diff(now(), State#state.ts)},
+  print_case_enter(TC, State, "Finished"),
+  {Return, State#state{ ts = undefined, tcs = [TCInfo | State#state.tcs] } }.
+
+%% @doc Called after post_init_per_suite, post_end_per_suite, post_init_per_group,
+%% post_end_per_group and post_end_per_testcase if the suite, group or test case failed.
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Line 92 is too long: %% post_end_per_group and post_end_per_testcase if the suite, group or test case failed..

File: src/ct_tty_hook.erl (hunk 11 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function pre_init_per_testcase/3

File: src/ct_tty_hook.erl [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+  print_case_enter(TC, State, "Starting"),
+  {Config, State#state{ ts = now(), total = State#state.suite_total + 1 } }.
+
+%% @doc Called after each test case.
+post_end_per_testcase(TC,_Config,Return,State) ->
+  TCInfo = {testcase, TC, Return, timer:now_diff(now(), State#state.ts)},
+  print_case_enter(TC, State, "Finished"),
+  {Return, State#state{ ts = undefined, tcs = [TCInfo | State#state.tcs] } }.
+
+%% @doc Called after post_init_per_suite, post_end_per_suite, post_init_per_group,
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Line 91 is too long: %% @doc Called after post_init_per_suite, post_end_per_suite, post_init_per_group,.

File: src/ct_tty_hook.erl (hunk 8 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function post_end_per_suite/4

File: src/ct_tty_hook.erl [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+%% post_end_per_group and post_end_per_testcase if the suite, group or test case failed.
+on_tc_fail(_TC, _Reason, State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function on_tc_fail/3

File: src/ct_tty_hook.erl [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+-export([post_end_per_testcase/4]).
+
+-export([on_tc_fail/3]).
+-export([on_tc_skip/3]).
+
+-export([terminate/1]).
+-record(state, { total, suite_total, ts, tcs, data, print_group, print_case }).
+
+%% @doc Return a unique id for this CTH.
+id(_Opts) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function id/1

File: src/ct_tty_hook.erl (hunk 6 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function post_init_per_group/4

File: src/ct_tty_hook.erl (hunk 5 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function pre_end_per_group/3

File: src/ct_tty_hook.erl [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+%% or due to an init function failing.
+on_tc_skip(_TC, Reason, State) ->
+  case Reason of
+    %% Just quit if files were not build correctly
+    {tc_user_skip, "Make failed"} -> erlang:halt(1);
+    _                             -> State
+  end.
+
+%% @doc Called when the scope of the CTH is done
+terminate(State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function terminate/1

File: src/ct_tty_hook.erl [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+print_suite_ok(Name, OkCount) ->
+  io:format("~n====== Suite OK: ~p (All ~p tests passed)~n",
+    [Name, OkCount]).
+print_suite_failed(Name, OkCount, FailCount, Cases) ->
+  ct:pal("~n====== Suite FAILED: ~p (~p of ~p tests failed)~n",
+    [Name, FailCount, OkCount+FailCount]),
+  lists:foreach(fun maybe_print_test_case/1, Cases).
+
+maybe_print_test_case({testcase, _Name,ok,_})              -> ok;
+maybe_print_test_case({testcase, Name,{error, Content},_}) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 134

File: src/ct_tty_hook.erl [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+  {Return, State#state{ ts = undefined, tcs = [TCInfo | State#state.tcs] } }.
+
+%% @doc Called after post_init_per_suite, post_end_per_suite, post_init_per_group,
+%% post_end_per_group and post_end_per_testcase if the suite, group or test case failed.
+on_tc_fail(_TC, _Reason, State) ->
+  State.
+
+%% @doc Called when a test case is skipped by either user action
+%% or due to an init function failing.
+on_tc_skip(_TC, Reason, State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function on_tc_skip/3

File: src/ct_tty_hook.erl (hunk 15 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 63

File: src/ct_tty_hook.erl (hunk 17 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 86

File: src/ct_tty_hook.erl [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+    State#state.data),
+  ok.
+
+
+aggregate_results(Cases) ->
+  Oks =   [ C || {testcase, _, ok, _} =         C <- Cases ],
+  Fails = [ C || {testcase, _, {error, _}, _} = C <- Cases ],
+  {length(Oks), length(Fails)}.
+
+print_suite({suites, Name,_, TestCases}) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Missing space after "," on line 119

File: src/ct_tty_hook.erl (hunk 9 above)
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: variable 'Suite' is unused

File: src/ct_tty_hook.erl [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+post_init_per_suite(_Suite,_Config,Return,State) ->
+  {Return, State}.
+
+%% @doc Called before end_per_suite.
+pre_end_per_suite(_Suite,Config,State) ->
+  {Config, State}.
+
+%% @doc Called after end_per_suite.
+post_end_per_suite(Suite,_Config,Return,State) ->
+  Data = {suites, Suite, State#state.suite_total, lists:reverse(State#state.tcs)},
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Elvis**:
> Line 58 is too long:   Data = {suites, Suite, State#state.suite_total, lists:reverse(State#state.tcs)},.

File: src/ct_tty_hook.erl [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+
+%% @doc Called after each end_per_group.
+post_end_per_group(Group,_Config,Return,State) ->
+  print_group_enter(Group, State, "Finished"),
+  {Return, State}.
+
+%% @doc Called before each test case.
+pre_init_per_testcase(TC,Config,State) ->
+  print_case_enter(TC, State, "Starting"),
+  {Config, State#state{ ts = now(), total = State#state.suite_total + 1 } }.
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: erlang:now/0: Deprecated BIF. See the "Time and Time Correction in Erlang" chapter of the ERTS User's Guide for more information.

File: src/ct_tty_hook.erl [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+
+%% @doc Called after each test case.
+post_end_per_testcase(TC,_Config,Return,State) ->
+  TCInfo = {testcase, TC, Return, timer:now_diff(now(), State#state.ts)},
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: erlang:now/0: Deprecated BIF. See the "Time and Time Correction in Erlang" chapter of the ERTS User's Guide for more information.

File: src/ct_tty_hook.erl [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+post_init_per_suite(_Suite, _Config, Return, State) ->
+  {Return, State}.
+
+%% @doc Called before end_per_suite.
+-spec pre_end_per_suite(any(), any(), any()) -> tuple().
+pre_end_per_suite(_Suite, Config, State) ->
+  {Config, State}.
+
+%% @doc Called after end_per_suite.
+post_end_per_suite(Suite, _Config, Return, State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function post_end_per_suite/4

File: src/ct_tty_hook.erl [hunk 30]
```diff
@@ -0,0 +1,163 @@
...
+  {ok, #state{total = 0, data = [],
+    print_group = PrintGroup, print_case = PrintCase}}.
+
+%% @doc Called before init_per_suite is called.
+-spec pre_init_per_suite(any(), any(), any()) -> tuple().
+pre_init_per_suite(_, Config, State) ->
+  {Config, State#state{suite_total = 0, tcs = []}}.
+
+%% @doc Called after init_per_suite.
+post_init_per_suite(_Suite, _Config, Return, State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function post_init_per_suite/4

File: src/ct_tty_hook.erl [hunk 31]
```diff
@@ -0,0 +1,163 @@
...
+%% @doc Called after each end_per_group.
+post_end_per_group(Group, _Config, Return, State) ->
+  print_group_enter(Group, State, "Finished"),
+  {Return, State}.
+
+%% @doc Called before each test case.
+-spec pre_init_per_testcase(any(), any(), any()) -> tuple().
+pre_init_per_testcase(TC, Config, State) ->
+  print_case_enter(TC, State, "Starting"),
+  {Config, State#state{ts = now(), total =
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: erlang:now/0: Deprecated BIF. See the "Time and Time Correction in Erlang" chapter of the ERTS User's Guide for more information.

File: src/ct_tty_hook.erl [hunk 32]
```diff
@@ -0,0 +1,163 @@
...
+    total = State#state.total + State#state.suite_total}}.
+
+%% @doc Called before each init_per_group.
+-spec pre_init_per_group(any(), any(), any()) -> tuple().
+pre_init_per_group(Group, Config, State) ->
+  print_group_enter(Group, State, "Starting"),
+  {Config, State}.
+
+%% @doc Called after each init_per_group.
+post_init_per_group(_Group, _Config, Return, State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function post_init_per_group/4

File: src/ct_tty_hook.erl [hunk 33]
```diff
@@ -0,0 +1,163 @@
...
+  {Return, State}.
+
+%% @doc Called after each end_per_group.
+-spec pre_end_per_group(any(), any(), any()) -> tuple().
+pre_end_per_group(_Group, Config, State) ->
+  {Config, State}.
+
+%% @doc Called after each end_per_group.
+post_end_per_group(Group, _Config, Return, State) ->
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: missing specification for function post_end_per_group/4

File: src/ct_tty_hook.erl [hunk 34]
```diff
@@ -0,0 +1,163 @@
...
+-spec pre_init_per_testcase(any(), any(), any()) -> tuple().
+pre_init_per_testcase(TC, Config, State) ->
+  print_case_enter(TC, State, "Starting"),
+  {Config, State#state{ts = now(), total =
+  State#state.suite_total + 1}}.
+
+%% @doc Called after each test case.
+-spec post_end_per_testcase(any(), any(), any(), any()) -> tuple().
+post_end_per_testcase(TC, _Config, Return, State) ->
+  TCInfo = {testcase, TC, Return, timer:now_diff(now(), State#state.ts)},
```
Comment (gadgetci) [inline src/ct_tty_hook.erl]: According to **Compiler**:
> Warning: erlang:now/0: Deprecated BIF. See the "Time and Time Correction in Erlang" chapter of the ERTS User's Guide for more information.

**Pull request merged**